data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java

```diff
@@ -60,6 +61,8 @@ public KafkaBuffer(final PluginSetting pluginSetting, final KafkaBufferConfig ka
         this.shutdownInProgress = new AtomicBoolean(false);
         final List<KafkaCustomConsumer> consumers = kafkaCustomConsumerFactory.createConsumersForTopic(kafkaBufferConfig, kafkaBufferConfig.getTopic(),
             innerBuffer, pluginMetrics, acknowledgementSetManager, byteDecoder, shutdownInProgress);
+        emptyCheckingConsumers = kafkaCustomConsumerFactory.createConsumersForTopic(kafkaBufferConfig, kafkaBufferConfig.getTopic(),
```

Comment:
Why are these a different set of consumers? Can we use the existing consumers and call `isTopicEmpty` on them?

Reply:
Kafka doesn't allow multiple threads to access the same consumer. Since we are submitting the original set of consumers to an executor service, I couldn't find a way to reuse them without stopping the executor service.

I don't like this approach but I wasn't able to find a better solution. Hopefully I'm missing something and someone has a clever way to do this without duplication

Reply:
Yea, that is a good point. I suppose the alternative would be to have each existing consumer thread also perform the check for an empty topic and then write it. But, that could have performance issues. I'm ok with this approach considering that limitation.

Reply:
This is interesting hack, having a consumer in group, but not really subscribing to the topic to get partitions. I Initially I thought we are doubling the consumers, but looks like we subscribe to topic only in run() so emptyCheckingConsumers consumers are not really consuming any data just being used to fetch topic's control data. It's not very intuitive and I don't know what overhead these consumers will create in terms of memory/tcp connections etc. 

May be we should look into creating a new class that'll make use of AdminClient (https://kafka.apache.org/23/javadoc/index.html?org/apache/kafka/clients/admin/AdminClient.html) to fetch such metadata information w/o creating consumer/producers, which can also be re-used for other admin operations. I guess AdminClient should also be able to provide all the info on topic/consumer-group their committed offsets etc.

Another simpler way is to just rely on InnerBuffer (based on current implementation). if innerbuffer is empty (may be for some time period) and there is no partition re-balance in progress then assigned partitions are fully read because consumers are constantly polling for records and shoving them in inner buffer.

Reply:
I recall looking into the admin client to do this and deciding against it. It's been a few weeks and I don't remember what my logic was, maybe that it didn't support consumer metrics based on my search history.

I think we can revisit this if needed and see if an admin client approach would work. On paper I wouldn't think the idle consumers are eating too many resources but it's also unclear what all they might be doing under the hood.

Reply:
>Another simpler way is to just rely on current implementation of InnerBuffer. if innerbuffer is empty (may be for some time period) and there is no partition re-balance in progress then assigned partitions are fully read because consumers are constantly polling for records and shoving them in inner buffer.

I thought about this too but it seemed like there were enough edge cases that it would be difficult to have 100% confidence on emptiness. Especially with partition rebalancing, DataPrepper crashes, etc